source/encoding.c

```diff
@@ -0,0 +1,316 @@
+
+static const char *hex_chars = "0123456789abcdef";
+
+static const uint8_t base64_sentinal_value = 0xff;
+static const char base64_encoding_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static uint8_t base64_decoding_table[256];
+static int8_t base64_decoding_table_init = 0;
+
+static const uint8_t bytes_7 = 7 * 8;
+static const uint8_t bytes_6 = 6 * 8;
+static const uint8_t bytes_5 = 5 * 8;
+static const uint8_t bytes_4 = 4 * 8;
+static const uint8_t bytes_3 = 3 * 8;
+static const uint8_t bytes_2 = 2 * 8;
+static const uint8_t bytes_1 = 1 * 8;
+static const uint8_t mask = 0xff;
+
+
+int aws_hex_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
+
+    size_t encoded_len = (to_encode_len << 1) + 1;
+
+    if(encoded_len <= to_encode_len) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    if (!output) {
+        *output_size = encoded_len;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < encoded_len) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+    for (size_t i = 0; i < to_encode_len; ++i) {
+
+        output[written++] = hex_chars[to_encode[i] >> 4 & 0x0f];
+        output[written++] = hex_chars[to_encode[i] & 0x0f];
+    }
+
+    output[written++] = '\0';
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static uint8_t char_to_int(char character) {
+    if (character >= 'a' && character <= 'f') {
+        return (uint8_t)(10 + (character - 'a'));
+    }
+
+    if (character >= 'A' && character <= 'F') {
+        return (uint8_t)(10 + (character - 'A'));
+    }
+
+    return (uint8_t)(character - '0');
+}
+
+int aws_hex_decode(const char *to_decode, size_t to_decode_len, uint8_t *output, size_t *output_size) {
+    size_t decoded_length = (to_decode_len + 1) >> 1;
+
+    if (!output) {
+        *output_size = decoded_length;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < decoded_length) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+
+    size_t i = 0;
+
+    //if the buffer isn't even, prepend a 0 to the buffer.
+    if (to_decode_len & 0x01) {
+        i = 1;
+        output[written++] = char_to_int(to_decode[i]);
+    }
+
+    for (; i < to_decode_len; i += 2) {
+        uint8_t value = char_to_int(to_decode[i]) << 4;
+        value |= char_to_int(to_decode[i + 1]);
+        output[written++] = value;
+    }
+
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static size_t calculate_base64_decoded_length(const char *input, size_t len) {
+    if (len == 0) {
+        return 0;
+    }
+
+    size_t padding = 0;
+
+    if (len >= 2 && input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
+        padding = 2;
+    }
+    else if (input[len - 1] == '=') { /*last char is = */
+        padding = 1;
+    }
+
+    return (len * 3 / 4 - padding);
```

Comment:
What behavior do we expect for invalid input? Specifically input that has length not divisible by 4.

Note also that the `* 3` here may overflow.